pkg/startup/v3/startup_test.go

```diff
@@ -0,0 +1,92 @@
+package startup
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/openshift/openshift-azure/pkg/api"
+	"github.com/openshift/openshift-azure/pkg/cluster/names"
+	"github.com/openshift/openshift-azure/test/util/populate"
+	"github.com/sirupsen/logrus/hooks/test"
+)
+
+func getCS() *api.OpenShiftManagedCluster {
+	prepare := func(v reflect.Value) {
+		switch v.Interface().(type) {
+		case []api.IdentityProvider:
+			// set the Provider to AADIdentityProvider
+			v.Set(reflect.ValueOf([]api.IdentityProvider{{Provider: &api.AADIdentityProvider{Kind: "AADIdentityProvider"}}}))
+		}
+	}
+	var cs api.OpenShiftManagedCluster
+	populate.Walk(&cs, prepare)
+	cs.Config.ImageVersion = "311.123.456"
+	cs.Config.Images.PrometheusOperator = ":"
+	cs.Config.Images.PrometheusConfigReloader = ":"
+	cs.Config.Images.ConfigReloader = ":"
+	cs.Config.Images.Prometheus = ":"
+	cs.Config.Images.AlertManager = ":"
+	cs.Config.Images.Grafana = ":"
+	cs.Config.Images.OAuthProxy = ":"
+	cs.Config.Images.NodeExporter = ":"
+	cs.Config.Images.KubeStateMetrics = ":"
+	cs.Config.Images.KubeRbacProxy = ":"
+	cs.Properties.AgentPoolProfiles = []api.AgentPoolProfile{
+		{
+			Name:   string(api.AgentPoolProfileRoleMaster),
+			Count:  1,
+			VMSize: api.StandardD4sV3,
+			OSType: api.OSTypeLinux,
+			Role:   api.AgentPoolProfileRoleMaster,
+		},
+		{
+			Name:   string(api.AgentPoolProfileRoleCompute),
+			Count:  1,
+			VMSize: api.StandardD4sV3,
+			OSType: api.OSTypeLinux,
+			Role:   api.AgentPoolProfileRoleCompute,
+		},
+	}
+
+	return &cs
+}
+
+func TestGenerateMasterTemplates(t *testing.T) {
```

Comment:
we've collided, sorry.  I don't think tests are fundamentally adding anything that isn't in TestHashScaleSetStability, which has just merged.  That test cannot succeed if the templates don't generate correctly.  I think all the changes to pkg/startup/... in this PR can be reverted.